hazelcast/src/main/java/com/hazelcast/cache/impl/ICacheRecordStore.java

```diff
@@ -483,6 +483,15 @@
      */
     ObjectNamespace getObjectNamespace();
 
+    /**
+     * Merges the given {@link CacheMergeTypes} via the given {@link SplitBrainMergePolicy}.
+     *
+     * @param mergingEntry the {@link CacheMergeTypes} instance to merge
+     * @param mergePolicy  the {@link SplitBrainMergePolicy} instance to apply
+     * @return the used {@link CacheRecord} if merge is applied, otherwise {@code null}
+     */
+    CacheRecord merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Data, CacheMergeTypes> mergePolicy);
```

Comment:
My brain doesn't parse this anymore :D `merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Data, CacheMergeTypes> mergePolicy)` is already in the actual code.

I just tried to apply what you've said to `AtomicLong`:
```java
class MergeGreaterValueMergePolicy
    implements SplitBrainMergePolicy<Long, AtomicLongMergingValueImpl> {

    @Override
    public Long merge(AtomicLongMergingValueImpl mergingValue,
                      AtomicLongMergingValueImpl existingValue) {
        if (mergingValue.getValue() > existingValue.getValue()) {
            return mergingValue.getValue();
        }
        return existingValue.getValue();
    }

    @Override
    public void writeData(ObjectDataOutput out) {
    }

    @Override
    public void readData(ObjectDataInput in) {
    }
}
```
This doesn't pass the latest configuration check:
> com.hazelcast.config.ConfigurationException: The merge policy com.hazelcast.concurrent.atomiclong.AtomicLongSplitBrainTest$MergeGreaterValueMergePolicy can just be configured on data structures which provide the merging type com.hazelcast.spi.impl.merge.AtomicLongMergingValueImpl. See SplitBrainMergingTypes for supported merging types.

You can do something like this though:
```java
MergeGreaterValueMergePolicy
    implements SplitBrainMergePolicy<Long, MergingValue<Long>>
```

What am I missing?

Reply:
I was thinking on the level of the `ICacheRecordStore#merge` method contract - right now it requires that the parameter must be of type `SplitBrainMergePolicy<Data, CacheMergeTypes> mergePolicy` (for instance `class MergePolicy implements  SplitBrainMergePolicy<Data, CacheMergeTypes>`).
But the method signature doesn't allow covariance in the second generic type - the merge policy can't implement an interface where the second generic type is a subtype - like the one you provided, i.e. this doesn't pass the compiler:
```
SplitBrainMergePolicy<Data, CacheMergingEntryImpl> policy;
cache.merge(mergingEntry, policy);
```

It's the same like this - if you wanted a method that summed all numbers in a collection and wrote:
```
public static int sum(Collection<Number> numbers){
        return ...;
    }
```
This doesn't pass:
```
sum(new ArrayList<Integer>());
        sum(new ArrayList<Float>());
```
So you need to write:
```
public static int sum(Collection<? extends Number> numbers){
        return ...;
    }
```

But never mind, this is actually not necessary and the policy implementing a subtype in the generic parameter doesn't even pass the validation check - which is good. So you can ignore this comment and just keep it in mind for future work with generics.

Reply:
A `public CacheRecord merge(CacheMergeTypes mergingEntry, SplitBrainMergePolicy<Data, ? extends CacheMergeTypes> mergePolicy)` breaks some lines later, since (for some reason) `CacheMergeTypes` doesn't match `? extends CacheMergeTypes`.

So we decided to leave it as it for now.